// ngx-swiper-wrapper/esm5/lib/swiper.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SwiperComponent } from './swiper.component';
import { SwiperDirective } from './swiper.directive';
import * as ɵngcc0 from '@angular/core';
var SwiperModule = /** @class */ (function () {
    function SwiperModule() {
    }
SwiperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: SwiperModule });
SwiperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function SwiperModule_Factory(t) { return new (t || SwiperModule)(); }, imports: [[CommonModule], CommonModule] });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(SwiperModule, { declarations: function () { return [SwiperComponent, SwiperDirective]; }, imports: function () { return [CommonModule]; }, exports: function () { return [CommonModule, SwiperComponent, SwiperDirective]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(SwiperModule, [{
        type: NgModule,
        args: [{
                imports: [CommonModule],
                declarations: [SwiperComponent, SwiperDirective],
                exports: [CommonModule, SwiperComponent, SwiperDirective]
            }]
    }], function () { return []; }, null); })();
    return SwiperModule;
}());
export { SwiperModule };
